casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala

```diff
@@ -288,19 +287,19 @@ class RuntimeManagerImpl[F[_]: Concurrent: Metrics: Span: Log](
       startHash: StateHash,
       terms: Seq[InternalProcessedDeploy],
       replayDeploy: InternalProcessedDeploy => F[Option[ReplayFailure]]
-  ): F[Either[ReplayFailure, StateHash]] =
-    (for {
-      _ <- EitherT.right(runtime.replaySpace.reset(Blake2b256Hash.fromByteString(startHash)))
-      _ <- EitherT(
-            terms.tailRecM { ts =>
-              if (ts.isEmpty)
-                ().asRight[ReplayFailure].asRight[Seq[InternalProcessedDeploy]].pure[F]
-              else replayDeploy(ts.head).map(_.map(_.asLeft[Unit]).toRight(ts.tail))
-            }
-          )
-      _   <- EitherT.right(Span[F].mark("before-replay-deploys-create-checkpoint"))
-      res <- EitherT.right[ReplayFailure](runtime.replaySpace.createCheckpoint())
-    } yield res.root.toByteString).value
+  ): F[Either[ReplayFailure, StateHash]] = {
+    import cats.implicits.{catsStdInstancesForList, catsStdInstancesForOption}
+    runtime.replaySpace.reset(Blake2b256Hash.fromByteString(startHash)) >> terms.toList.traverse(
+      replayDeploy
+    ) >>= { replayedDeploys =>
+      replayedDeploys.foldK match {
+        case Some(replayError) => replayError.asLeft[StateHash].pure[F]
+        case None =>
+          Span[F].mark("before-replay-deploys-create-checkpoint") >>
+            runtime.replaySpace.createCheckpoint().map(_.root.toByteString.asRight[ReplayFailure])
+      }
```

Comment:
This is going to replay all deploys even if the first fails.

I think we should make/convert the result type of replayDeploy to be Either[ReplayError, Unit] and do a `terms.toList.foldM`

Also, I think the changes you did would be more readable if the for-comprehension stayed despite the EitherT going away :)

Reply:
See for example https://github.com/rchain/rchain/blob/2485709556f9d1d22db6aedd34e2ebcca6ab9e57/casper/src/main/scala/coop/rchain/casper/util/rholang/InterpreterUtil.scala#L214-L222

Reply:
Actually, isn't this method a third copy of 'replay block deploys' code?
See the comment at https://github.com/rchain/rchain/blob/2485709556f9d1d22db6aedd34e2ebcca6ab9e57/casper/src/main/scala/coop/rchain/casper/util/rholang/InterpreterUtil.scala#L66-L67